module_utils/network_lsr/argument_validator.py

```diff
@@ -1733,6 +1737,13 @@ def _validate_post(self, value, name, result):
     VALIDATE_ONE_MODE_INITSCRIPTS = "initscripts"
 
     def validate_connection_one(self, mode, connections, idx):
```

Comment:
@thom311 what is the purpose of `validate_connection_one()`? Is this the right place for this check? What does `one` relate to, here? Is it one provider?

Reply:
It "validate"s "one" "connection", while also taking into account the provider/mode.

It's noteworthy that it validates one connection (`connections[idx]`), because `ArgValidator_ListConnections` in general is about a list of connections.

The earlier validation steps done by ArgValidator are independent of the mode.

Reply:
Why are those separated? `validate_connection_one()` is only used in a loop to validate all connections.

Reply:
because the caller want's to know which of the connections was invalid. If `validate_connection_all()` would only throw an exception, it would need to somehow propagate `idx`. It seems better to have smaller, self-contained functions, that do something understandable. E.g. it's understandable that `validate_connection_one()` does either nothing or fail with a `ValidationError` on error. It's less clear how it would behave, when validating all connections at once.

But yes, it could also be different...

Reply:
There are already a lot of validations happening in `.validate()` before `validdate_connetion_one()` is called and there the index is reported in the exception message:
https://github.com/linux-system-roles/network/blob/main/module_utils/network_lsr/argument_validator.py#L1696

Reply:
`ArgValidator_ListConnections._validate_post()` checks interdependencies between profiles, so necessarily that processes the entire list of connection. On the other hand, `ArgValidator._validate_impl()` indeed loads (and validates) each element one-by-one, indpendently. But in general, the API or behavior of those two could not reasonably be any different and if they need to report the `idx`, then the should attach it to `ValidationError` (and kind of they already do, via the `name` and `error_message`).

The API cannot be reasonably different than what it is, is because ArgValidator should be nestable/composable, so there is no place where somebody would drive the creation/validation of individual list elements (otherwise, you'd have to know that the element you are creating is a `ArgValidatorList` type). Also, `.validate()` not only validates but normalizes (modifies) the data.

On the other hand, `validate_connection_one()` does not modify/normalize anything. Also, the caller is well aware that they want to specifically validate a connection of `ArgValidator_ListConnections`. It is not abstract "ArgValidator.validate_one_list_element()`, because the caller necessarily knows what `ArgValidator_ListConnections` and `validate_connection_one()` does. As such, the API can be different, and it is.


Of you dislike that, please open a merge request.